### PR TITLE
[Sprint 1] Main page-Databinding

### PR DESCRIPTION
**Pull request by nguyenminhduong**
Redmine https://dev.framgia.com/redmine/issues/112027
Databinding

### CODE REVIEW

File: app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/ItemProductViewModel.java [hunk 1]
```diff
@@ -17,4 +18,39 @@ public ItemProductViewModel(Product product,
...
+    }
+
+    public void onItemClicked(View view) {
+        if (mItemClickListener == null) {
+            return;
+        }
+        mItemClickListener.onItemRecyclerViewClick(mProduct);
+    }
+
+    public void addCart(View view) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/ItemProductViewModel.java]: move this method to MainPage Viewmodel
Write callback in here like [#comment](https://github.com/framgia/android_forder_01/pull/89/files#diff-29fed8c51a1389bd65301f012fdc33ebR34)

File: app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/MainPageFragment.java [hunk 2]
```diff
@@ -23,12 +26,15 @@ public static MainPageFragment newInstance() {
 
     public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
             @Nullable Bundle savedInstanceState) {
-        mViewModel = new MainPageViewModel();
+        ArrayList<Product> products = new ArrayList<>();
+        ArrayList<Shop> shops = new ArrayList<>();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/MainPageFragment.java]: Change to
```java
List<..> abcs = new ArrayList<>()
```

File: app/src/main/java/com/framgia/forder/utils/binding/BindingUtils.java [hunk 3]
```diff
@@ -35,4 +37,9 @@ public static void setViewPagerAdapter(final ViewPager viewPager,
     public static void setViewPagerTabs(final TabLayout tabLayout, final ViewPager viewPager) {
         tabLayout.setupWithViewPager(viewPager, true);
     }
+
+    @BindingAdapter("bind:imageUrl")
+    public static void loadImage(ImageView imageView, String url) {
+        Glide.with(imageView.getContext()).load(url).into(imageView);
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/forder/utils/binding/BindingUtils.java]: add a placeholder will be better

File: app/src/main/java/com/framgia/forder/data/model/Product.java [hunk 4]
```diff
@@ -154,7 +156,16 @@ public void setStatus(String status) {
...
+    public String getFormatStartHour() {
+        return mStartHour.substring(START_INDEX, END_INDEX);
+    }
+
+    public String getFormatEndHour() {
+        return mEndHour.substring(START_INDEX, END_INDEX);
+    }
+
     @Override
+
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/model/Product.java]: remove empty line

File: app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/ItemShopViewModel.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package com.framgia.forder.screen.mainpagetemp.mainpage.mainpagetab;
 
 import android.databinding.BaseObservable;
+import android.view.View;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/ItemShopViewModel.java]: unuse import,
Please format code before commit

File: app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/MainPageViewModel.java [hunk 6]
```diff
@@ -1,14 +1,31 @@
...
+    private ProductAdapter mProductAdapter;
+    private ShopAdapter mShopAdapter;
+
+    public MainPageViewModel(ProductAdapter productAdapter, ShopAdapter shopAdapter) {
+        mProductAdapter = productAdapter;
+        mShopAdapter = shopAdapter;
+    }
 
-    public MainPageViewModel() {
+    public ProductAdapter getProductAdapter() {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/MainPageViewModel.java]: please rearrange method in class
``1. ``override`` -> 2 ``private`` 3. ``public``

File: app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/MainPageViewModel.java [hunk 7]
```diff
@@ -25,4 +42,13 @@ public void onStop() {
     public void setPresenter(MainPageContract.Presenter presenter) {
         mPresenter = presenter;
     }
+
+    @Override
+    public void onItemRecyclerViewClick(Product item) {
+
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/MainPageViewModel.java]: add ``// TODO implement later`` in here

File: app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/MainPageViewModel.java [hunk 8]
```diff
@@ -25,4 +42,13 @@ public void onStop() {
...
+    }
+
+    public void addCart(Product product) {
+        // todo add card
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/MainPageViewModel.java]: change to  ``// TODO implement later`` in here

File: app/src/main/res/layout/item_shop.xml [hunk 9]
```diff
@@ -1,9 +1,18 @@
 <?xml version="1.0" encoding="utf-8"?>
 <layout
-    xmlns:android="http://schemas.android.com/apk/res/android">
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    >
+    <data>
+        <variable
+            name="viewModel"
+            type="com.framgia.forder.screen.mainpagetemp.mainpage.mainpagetab.ItemShopViewModel"
```
Comment (daolq3012) [inline app/src/main/res/layout/item_shop.xml]: change package tree

Comment (nguyenminhduong) [inline app/src/main/res/layout/item_shop.xml]: Chổ này change như thế nào anh, do chưa có home page nên em chưa bỏ vào được

Comment (daolq3012) [inline app/src/main/res/layout/item_shop.xml]: mang máy sang anh bày e ei

File: app/src/main/java/com/framgia/forder/data/model/Product.java [hunk 10]
```diff
@@ -154,6 +158,14 @@ public void setStatus(String status) {
         mStatus = status;
     }
 
+    public String getFormatStartHour() {
+        return convertUiFormatToDataFormat(mStartHour, INPUT_TIME_FORMAT, OUTPUT_TIME_FORMAT);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/model/Product.java]: should be ``Utils.DateTimeUntils.convertUiFormatToDataFormat (...)``  It more clearly

File: app/src/main/java/com/framgia/forder/data/model/Product.java [hunk 11]
```diff
@@ -154,6 +158,14 @@ public void setStatus(String status) {
...
+    }
+
+    public String getFormatEndHour() {
+        return convertUiFormatToDataFormat(mEndHour, INPUT_TIME_FORMAT, OUTPUT_TIME_FORMAT);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/model/Product.java]: should be ``Utils.DateTimeUntils.convertUiFormatToDataFormat (...)``  It more clearly

File: app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/ItemProductViewModel.java [hunk 12]
```diff
@@ -11,10 +11,51 @@
 public class ItemProductViewModel extends BaseObservable {
     private Product mProduct;
     private BaseRecyclerViewAdapter.OnRecyclerViewItemClickListener<Product> mItemClickListener;
+    private MainPageViewModel mListener;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpagetemp/mainpage/mainpagetab/ItemProductViewModel.java]: remove

File: app/src/main/AndroidManifest.xml [hunk 13]
```diff
@@ -25,7 +25,7 @@
             android:theme="@style/AppTheme"/>
         <!-- TODO remove later -->
         <activity
-            android:name=".screen.mainpagetemp.MainPageTempActivity"
+            android:name=".screen.mainpage.MainPageTempActivity"
```
Comment (daolq3012) [inline app/src/main/AndroidManifest.xml]: remove this activity, you can inject ``MainPageContainerFragment.java`` into MainActivity

File: app/src/main/java/com/framgia/forder/screen/mainpage/MainPageViewModel.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+import com.framgia.forder.data.model.Product;
+import com.framgia.forder.screen.BaseRecyclerViewAdapter;
+import java.util.Observable;
+
+/**
+ * Exposes the data to be used in the Main screen.
+ */
+
+interface OrderListener {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpage/MainPageViewModel.java]: OrderListener should be inside class MainPageViewModel

File: app/src/main/java/com/framgia/forder/screen/mainpage/MainPageViewModel.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public void addCart(Product product) {
+        // TODO implement later in here
+    }
+}
+
+;
+
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpage/MainPageViewModel.java]: (tat)(2tat)

File: app/src/main/java/com/framgia/forder/screen/main/FragmentContainer.java [hunk 16]
```diff
@@ -38,7 +39,8 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
         @MainViewModel.Tab int tab = getArguments().getInt(EXTRA_TAB_FOOTER);
         switch (tab) {
             case TAB_HOME:
-                //TODO new Fragment Home
+                navigator.goNextChildFragment(R.id.layout_content,
+                        MainPageContainerFragment.newInstance(), true, 2, "MainPage");
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/main/FragmentContainer.java]: change to
```java
navigator.goNextChildFragment(R.id.layout_content,
                         MainPageContainerFragment.newInstance(), true, Navigator.NavigateAnim.NONE, "MainPageContainerFragment");
```

File: app/src/main/java/com/framgia/forder/screen/mainpage/shop/ShopAdapter.java [hunk 17]
```diff
@@ -44,23 +45,24 @@ public int getItemCount() {
         return mShops.size();
     }
 
-    void setItemClickListener(OnRecyclerViewItemClickListener<Shop> itemClickListener) {
+    public void setItemClickListener(MainPageViewModel itemClickListener) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpage/shop/ShopAdapter.java]: change to ```setItemClickListener(OnRecyclerViewItemClickListener<Object> itemClickListener)

File: app/src/main/java/com/framgia/forder/screen/mainpage/product/OrderListener.java [hunk 18]
```diff
@@ -0,0 +1,11 @@
+package com.framgia.forder.screen.mainpage.product;
+
+import com.framgia.forder.data.model.Product;
+
+/**
+ * Created by Duong on 4/14/2017.
+ */
+
+public interface OrderListener {
+    void addCart(Product product);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/mainpage/product/OrderListener.java]: change to ``onAddToCart(Product product)``

**Pull request merged**